homework04/breakout.py

```diff
@@ -0,0 +1,228 @@
+    x = tf.keras.layers.Conv2D(filters=64, kernel_size=3, activation='relu', padding='same')(x)
+    x = tf.keras.layers.Conv2D(filters=64, kernel_size=3, activation='relu', padding='same')(x) + x # residual connection
+    x = tf.keras.layers.Conv2D(filters=64, kernel_size=3, activation='relu', padding='same')(x) + x # residual connection
+    x = tf.keras.layers.MaxPool2D(pool_size=2)(x)
+    x = tf.keras.layers.Conv2D(filters=64, kernel_size=3, activation='relu', padding='same')(x) + x # residual connection
+    x = tf.keras.layers.Conv2D(filters=64, kernel_size=3, activation='relu', padding='same')(x) + x # residual connection
+    x = tf.keras.layers.Conv2D(filters=64, kernel_size=3, activation='relu', padding='same')(x) + x # residual connection
+    x = tf.keras.layers.GlobalAveragePooling2D()(x)
+    x = tf.keras.layers.Dense(units=64, activation='linear')(x) + x # residual connection
+    x = tf.keras.layers.Dense(units=num_actions, activation='linear')(x)
+    model = tf.keras.Model(inputs=input_layer, outputs=x)
+    return model
+
+def train_dqn(train_dqn_network, target_network, dataset, optimizer, gamma: float, num_training_steps: int, batch_size: int=256):
+    dataset = dataset.batch(batch_size).prefetch(4)
+    @tf.function
+    def training_step(q_target, observations, actions):
+        with tf.GradientTape() as tape:
+            q_predictions_all_actions = train_dqn_network(observations) # shape: (batch_size, num_actions)
+            q_predictions = tf.gather(q_predictions_all_actions, actions, batch_dims=1)
+            loss = tf.reduce_mean(tf.square(q_predictions - q_target))
+        gradients = tape.gradient(loss, train_dqn_network.trainable_variables)
+        optimizer.apply_gradients(zip(gradients, train_dqn_network.trainable_variables))
+        return loss
+    
+    losses = []
+    q_values = []
+    for i, state_transition in enumerate(dataset):
+        state, action, reward, subsequent_state, terminated = state_transition
+        q_vals = target_network(subsequent_state)
+        q_values.append(q_vals.numpy())
+        max_q_values = tf.reduce_max(q_vals, axis=1)
+        use_subsequent_state = tf.where(terminated, tf.zeros_like(max_q_values, dtype=tf.float32), tf.ones_like(max_q_values, dtype=tf.float32))
+        q_target = reward + (gamma * max_q_values * use_subsequent_state)
+        loss = training_step(q_target, observations=state, actions=action)
+        losses.append(loss)
+        if i >= num_training_steps:
+            break
+    return np.mean(losses), np.mean(q_values)
+
+def test_q_network(test_dqn_network, environment_name: str, num_parallel_tests: int, gamma: float):
+    envs = gym.vector.make(environment_name, num_parallel_tests)
+    states, _ = envs.reset()
+    done = False
+    timestep = 0
+    episodes_finished = np.zeros(num_parallel_tests, dtype=np.bool)
+    returns = np.zeros(num_parallel_tests)
+
+    while not done:
+        q_values = test_dqn_network(states)
```

Comment:
with this version I get the following error:
```
ValueError: Input 0 of layer "model" is incompatible with the layer: expected shape=(None, 84, 84, 3), found shape=(10, 210, 160, 3)
```

based on our discussion I made the following change and was able to run it, so let's change it here too:
```suggestion
        q_values = test_dqn_network(observation_preprocessing_function(states))
```